### PR TITLE
Add 2 DDB Streams workers

### PR DESCRIPTION
**Pull request by eric-wu**

### CODE REVIEW

File: src/main/java/org/sagebionetworks/bridge/workers/App.java [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+    }
+
+    private static void startWorker(final Worker worker, final AnnotationConfigApplicationContext ctx) {
+        ctx.addApplicationListener(new ApplicationListener<ContextClosedEvent>() {
+            @Override
+            public void onApplicationEvent(ContextClosedEvent event) {
+                worker.shutdown();
+            }
+        });
+        worker.run();
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/App.java]: Does this launch the worker synchronously or asynchronously? Do we need an Executor to handle multithreading?

File: src/main/java/org/sagebionetworks/bridge/workers/BridgeWorkersConfig.java [hunk 2]
```diff
@@ -4,50 +4,217 @@
...
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.InitialPositionInStream;
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.KinesisClientLibConfiguration;
+import com.amazonaws.services.kinesis.clientlibrary.lib.worker.Worker;
 import com.google.common.base.StandardSystemProperty;
 
 @Configuration
 public class BridgeWorkersConfig {
 
-    @Bean(name = "config")
+    private final Logger log = LoggerFactory.getLogger(App.class);
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/BridgeWorkersConfig.java]: Nitpick: BridgeWorkersConfig.class

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/DynamoStreams.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+    private final Map<String, DynamoStream> streams;
+
+    public DynamoStreams(final List<String> fqTableNames, final AmazonDynamoDB dynamo) {
+        streams = new HashMap<>();
+        StreamsUtils.getStreams(fqTableNames, dynamo).stream().forEach(
+                stream -> streams.put(stream.getTableName(), stream));
+    }
+
+    public List<DynamoStream> getStreams() {
+        return Collections.unmodifiableList(new ArrayList<DynamoStream>(streams.values()));
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/DynamoStreams.java]: I like Guava's ImmutableList for this. You can shorten this to ImmutableList.copyOf(streams.values());

Another advantage of Guava ImmutableList is that, unlike Collections.unmodifiableList(), the ImmutableList is an immutable copy, while Collections.unmodifiableList() is a view into a collection that could still be modified from under it.

File: src/main/java/org/sagebionetworks/bridge/workers/App.java (hunk 1 above)
Comment (eric-wu) [inline src/main/java/org/sagebionetworks/bridge/workers/App.java]: Asynchronously. It uses a cached thread pool by default if you don't specify an executor.

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+                log().info("[EventId=" + streamRecord.getEventID() + ", " +
+                        "SequenceNumber=" + streamRecord.getDynamodb().getSequenceNumber() + "]");
+                process(streamRecord);
+                count += 1;
+                if (count % checkpointInterval == 0) {
+                    try {
+                        checkpointer.checkpoint();
+                    } catch(Exception e) {
+                        throw new RuntimeException(e);
+                    }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java]: Instead of catching all Exceptions, we should only catch the exceptions checkpointer.checkpoint() declares (or RuntimeException if it doesn't declare any).

Similarly, instead of throwing RuntimeException (which in tests, can easily be mixed up with a NullPointerException due to a bug in the code), we should throw a more specific exception or create our own RuntimeException to throw.

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+    }
+
+    @Override
+    public void shutdown(IRecordProcessorCheckpointer checkpointer, ShutdownReason reason) {
+        if(reason == ShutdownReason.TERMINATE) {
+            try {
+                checkpointer.checkpoint();
+            } catch (Exception e) {
+                throw new RuntimeException(e);
+            }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java]: Same here.

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+                break;
+            case "MODIFY":
+                onModify(streamRecord);
+                break;
+            case "REMOVE":
+                onRemove(streamRecord);
+                break;
+            default:
+                throw new RuntimeException(
+                        "Unexpected event name: " + streamRecord.getEventName());
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessor.java]: Similarly, we should throw a more specific exception here.

File: src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessorTest.java [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+public class StreamRecordProcessorTest {
+
+    private final StreamRecordProcessor processor;
+
+    public StreamRecordProcessorTest() {
+        processor = new StreamRecordProcessor(2) {
+            @Override
+            Logger log() {
+                return mock(Logger.class);
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessorTest.java]: Would it be better to use a real logger? That way, if unit tests fail, we can use the logs to help figure out what happened.

File: src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessorTest.java [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+import com.amazonaws.services.dynamodbv2.model.Record;
+import com.amazonaws.services.dynamodbv2.model.StreamRecord;
+import com.amazonaws.services.dynamodbv2.streamsadapter.model.RecordAdapter;
+import com.amazonaws.services.kinesis.clientlibrary.interfaces.IRecordProcessorCheckpointer;
+
+public class StreamRecordProcessorTest {
+
+    private final StreamRecordProcessor processor;
+
+    public StreamRecordProcessorTest() {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamRecordProcessorTest.java]: Minor: Using a constructor for JUnit is atypical. Maybe move this to a @Before?

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtils.java [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+    public static List<DynamoStream> getStreams(final List<String> fqTableNames, final AmazonDynamoDB dynamo) {
+        checkNotNull(fqTableNames);
+        checkNotNull(dynamo);
+        final List<DynamoStream> streams = new ArrayList<>();
+        for (final String fqTableName : fqTableNames) {
+            if (!isStreamEnabled(fqTableName, dynamo)) {
+                LOG.info("Enabling stream for table " + fqTableName + "...");
+                enableStream(fqTableName, dynamo);
+                LOG.info("Stream enabled for table " + fqTableName + "... Done.");
+            }
```
Comment (DwayneJengSage) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtils.java]: This is called during Spring Initialization. What happens if a new table is created while this is running? Will we have to restart our workers? Similarly with getReplicaTables().

File: src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,5 @@
+package org.sagebionetworks.bridge.workers.dynamodb.streams;
+
+public class StreamsUtilsTest {
```
Comment (DwayneJengSage) [inline src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtilsTest.java]: Empty test file?

It might make it easier to test StreamsUtils if it were non-static and there were a non-static wrapper around DynamoUtils.getExistingTables() so you can mock it.

File: src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtils.java (hunk 9 above)
Comment (eric-wu) [inline src/main/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtils.java]: We will have to restart the workers with this design.

File: src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtilsTest.java (hunk 10 above)
Comment (eric-wu) [inline src/test/java/org/sagebionetworks/bridge/workers/dynamodb/streams/StreamsUtilsTest.java]: This shouldn't be checked in.  Mocking around is expensive.  That's why it's empty.  But StreamsUtils has been manually tested.
